src/theory/bv/bvgauss.cpp

```diff
@@ -0,0 +1,715 @@
+/*********************                                                        */
+/*! \file bvgauss.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Aina Niemetz
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Gaussian Elimination preprocessing pass.
+ **
+ ** Simplify a given equation system modulo a (prime) number via Gaussian
+ ** Elimination if possible.
+ **/
+
+#include "theory/bv/bvgauss.h"
+
+#include <stack>
+#include <unordered_map>
+
+#include "theory/rewriter.h"
+#include "theory/bv/theory_bv_utils.h"
+#include "theory/bv/theory_bv_rewrite_rules_normalization.h"
+
+using namespace CVC4;
+using namespace std;
+
+namespace CVC4 {
+namespace theory {
+namespace bv {
+
+static bool is_bv_const(Node n)
+{
+  if (n.isConst()) { return true; }
+  return Rewriter::rewrite(n).getKind() == kind::CONST_BITVECTOR;
+}
+
+static Node get_bv_const(Node n)
+{
+  Assert(is_bv_const(n));
+  return Rewriter::rewrite(n);
+}
+
+static Integer get_bv_const_value(Node n)
+{
+  Assert(is_bv_const(n));
+  return get_bv_const(n).getConst<BitVector>().getValue();
+}
+
+/* Note: getMinBwExpr assumes that 'expr' is rewritten.
+ *
+ * If not, all operators that are removed via rewriting (e.g., ror, rol, ...)
+ * will be handled via the default case, which is not incorrect but also not
+ * necessarily the minimum. */
+unsigned BVGaussElim::getMinBwExpr(Node expr)
+{
+  stack<Node> visit;
+  /* Maps visited nodes to the determined minimum bit-width required. */
+  unordered_map<Node, unsigned, NodeHashFunction> visited;
+  unordered_map<Node, unsigned, NodeHashFunction>::iterator it;
+
+  visit.push(expr);
+  while (!visit.empty())
+  {
+    Node n = visit.top();
+    visit.pop();
+    it = visited.find(n);
+    if (it == visited.end())
+    {
+      if (is_bv_const(n))
+      {
+        /* Rewrite const expr, overflows in consts are irrelevant. */
+        visited[n] = get_bv_const(n).getConst<BitVector>().getValue().length();
+      }
+      else
+      {
+        visited[n] = 0;
+        visit.push(n);
+        for (const Node &nn : n) { visit.push(nn); }
+      }
+    }
+    else if (it->second == 0)
+    {
+      Kind k = n.getKind();
+      Assert(k != kind::CONST_BITVECTOR);
+      Assert(!is_bv_const(n));
+      switch (k)
+      {
+        case kind::BITVECTOR_EXTRACT:
+        {
+          unsigned w = utils::getSize(n);
+          visited[n] = std::min(
+              w, std::max(visited[n[0]] - utils::getExtractLow(n), 0u));
```

Comment:
Since you are using `using namespace std;` the std:: here are redundant. Either always use std:: (and not using namespace std;) or remove redundant namespace prefix.